bin/start_sync_stop.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+"""A script to test rippled in an infinite loop of start-sync-stop.
+
+- Expects 'rippled' to be on the `PATH` and 'rippled.cfg' to be in the current
+  directory.
+- Takes no arguments.
+- Can be stopped with SIGINT.
+- Tested on Python 3.8.
+- Has no dependencies outside the standard library.
+"""
+
+import asyncio
+import contextlib
+import json
+import logging
+import os
+import subprocess
+import time
+import urllib.error
+import urllib.request
+
+
+@contextlib.asynccontextmanager
+async def rippled(path='rippled', config='rippled.cfg'):
+    """A context manager for a rippled process."""
+    # Start the server.
+    process = await asyncio.create_subprocess_exec(
+        path,
+        '--conf',
+        config,
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+    )
+    logging.info(f'rippled started with pid {process.pid}')
+    try:
+        yield process
+    finally:
+        # Ask it to stop.
+        logging.info(f'asking rippled ({process.pid}) to stop')
+        start = time.time()
+        process.terminate()
+
+        # Wait nicely.
+        try:
+            await asyncio.wait_for(process.wait(), 60)
+        except asyncio.TimeoutError:
+            # Ask the operating system to kill it.
+            logging.warning(f'killing rippled ({process.pid})')
+            try:
+                process.kill()
+            except ProcessLookupError:
+                pass
+
+        code = await process.wait()
+        end = time.time()
+        logging.info(
+            f'rippled stopped after {end - start} seconds with code {code}'
+        )
+
+
+async def sync(port, duration=60, interval=5):
+    """Poll rippled on an interval until it has been synced for a duration."""
+    stopwatch = 0
+    while stopwatch < duration:
+        await asyncio.sleep(interval)
+
+        request = urllib.request.Request(
+            f'http://127.0.0.1:{port}',
+            data=json.dumps({
+                'method': 'server_state'
+            }).encode(),
+            headers={'Content-Type': 'application/json'},
+        )
+        with urllib.request.urlopen(request) as response:
+            try:
+                body = json.loads(response.read())
+            except urllib.error.HTTPError as cause:
+                logging.warning(f'server_state returned s.o.t. JSON: {cause}')
+                stopwatch = 0
+                continue
+
+        try:
+            state = body['result']['state']['server_state']
+        except KeyError as cause:
+            logging.warning(f'server_state response missing key: {cause.key}')
+            stopwatch = 0
+            continue
+        logging.info(f'server_state: {state}')
+        if state == 'full':
```

Comment:
There are three different state values that all correspond to `full`: `full`, `validating`, and `proposing`.  You should allow all three.

I also suggest that you require that the server _stay_ in sync for the full duration.  What you currently have allows a server to fall out of sync repeatedly.  I think you should require the fully synced time to be contiguous.

Given that, I suggest you change the code to something like this:
```
        full_states = {'full', 'validating', 'proposing'}
        if state in full_states:
            stopwatch += interval
        else:
            # rippled must stay in sync for the full duration.  Start over if
            # it loses sync.
            stopwatch = 0
```